### ISSUE TITLE
Add extractor for data from EnvironmentLogger --> geostreams database

### ISSUE THREAD

**Issue by dlebauer** [terraref/computing-pipeline#252]
* The data are transferred outside the core lemnatec pipeline -> JD, where are these?
* Otherwise the conversion should be similar.

**Comment by Zodiase** [comment]
@dlebauer Who would be the person that knows where the data are and what is the schema/format of them?

**Comment by jdmaloney** [comment]
@Zodiase If you have access to Roger or the Terraref Globus Endpoint, the files are located at:
/gpfs/largeblockFS/projects/arpae/terraref/sites/ua-mac/raw_data/weather --> from Roger perspective
or
/gpfs/largeblockFS/projects/arpae/terraref/sites/ua-mac/raw_data/weather --> Globus perspective

**Comment by Zodiase** [comment]
@dlebauer @jdmaloney I roughly scanned the data files and noticed an inconsistency in file size/data point count and in some cases the naming conventions.

For example, `2017-01-26/WeatherStation_SecData_2017_01_26_1802.dat` is more than 10 times the size of `2017-01-27/WeatherStation_SecData_2017_01_27_0915.dat`. And while most folders contain files named `WeatherStation_SecData_*_*_*_*.dat`, some folders only contain one file `WeatherStation_SecData.dat`, but then there is file `2016-08-09/WeatherStation_SecData.dat` sitting next to `2016-08-09/WeatherStation_SecData_2016_08_09_2234.dat` and some others.

Is there a pattern to follow here? For the previous met extractor I wrote, it relies on the datasets having 24 partitioned `.dat` files containing non-overlapping data.

**Comment by jdmaloney** [comment]
@Zodiase I think there was a change in naming format there.  2016_08_09 was early on in the data collection and I think its naming scheme should be ignored, it's only for those first few days (new schema starts at 2016_08_10) and has been consistent since then.  In terms of the size variance I am not sure.  I'm just grabbing the files off the Network Storage device they have down at Maricopa, how the files are generated and what goes into them I do not know.

**Comment by dlebauer** [comment]
@Andrade-Pedro can you comment on this? I believe this is the second met station you set up. Also can you give approximate coordinates

**Comment by dlebauer** [comment]
@Andrade-Pedro I think JD answered but coordinates would be nice to have

**Comment by Andrade-Pedro** [comment]
Ground weather station coordinates: 33.074457 N, 111.975163 W

**Comment by Zodiase** [comment]
@jdmaloney Wait I just noticed the data files you pointed to me has already been handled by the first Met extractor in #156. The path you gave me is exactly the same as the one in this comment: https://github.com/terraref/computing-pipeline/issues/156#issuecomment-248984197

**Comment by max-zilla** [comment]
@jdmaloney what data is David talking about "transferred outside core lemnatec pipeline"? We move the weather station data via pipeline. Does he mean the lightning data that we're doing the append-to-single-file stuff?

**Comment by dlebauer** [comment]
Max there is the Environmental Logger and the Weather Station. There was
some confusion initially - Xingchen had already processed the weather
station data and will now work on the Environmental Logger data.

**Comment by jdmaloney** [comment]
@Zodiase Here is the path to the other data:

/gpfs/largeblockFS/projects/arpae/terraref/sites/ua-mac/raw_data/EnvironmentLogger

**Comment by Zodiase** [comment]
# Source File Schema

## Document (with example values)

```JSON
{
  "environment_sensor_fixed_infos": {
    "par_sensor": {
      "manufacturer": "www.apogeeinstruments.com",
      "model": "SQ214",
      "location in gantry system": "top of gantry"
    },
    "co2_sensor": {
      "sensor manufacturer": "Vaisala",
      "model": "Carbocap CO2 Probe GMP343 A1C1B0N0N0B",
      "sensor serial number": "L3420008",
      "additional info": "SO 5530060878",
      "calbration date": "2015.08.18",
      "location in gantry system": "camera box",
      "analog digital interface": "WAGO 750-478"
    },
    "weather_station": {
      "manufacturer": "www.thiesclima.com",
      "article nr": "4.9200.00.000",
      "analog digital interface": "wago 750-478",
      "location in gantry system": "top of gantry"
    },
    "spectrometer": {
      "manufacturer": "www.oceanoptics.com",
      "model": "STS-VIS",
      "location in gantry system": "top of gantry"
    }
  },
  "environment_sensor_readings": Array.<SensorReading>
}
```

## SensorReading

```JSON
{
  "timestamp": "2017.01.10-00:32:02",
  "weather_station": {
    "sunDirection": {
      "value": "357.5829340495",
      "unit": "degrees",
      "rawValue": "9.93285927915281"
    },
    "airPressure": {
      "value": "1017.8899502548",
      "unit": "hPa",
      "rawValue": "8.38038270210883"
    },
    "brightness": {
      "value": "1.0364391003",
      "unit": "kilo Lux",
      "rawValue": "0.00183111056855983"
    },
    "relHumidity": {
      "value": "75.0328073977",
      "unit": "relHumPerCent",
      "rawValue": "7.50328073976867"
    },
    "temperature": {
      "value": "11.542710654",
      "unit": "DegCelsius",
      "rawValue": "5.15427106540117"
    },
    "windDirection": {
      "value": "315.9984130375",
      "unit": "degrees",
      "rawValue": "8.77773369548631"
    },
    "precipitation": {
      "value": "0.0406736656",
      "unit": "mm/h",
      "rawValue": "0.00396740623187964"
    },
    "windVelocity": {
      "value": "1.5326395459",
      "unit": "m/s",
      "rawValue": "0.255439924314097"
    }
  },
  "sensor par": {
    "value": "0",
    "unit": "umol/(m^2*s)",
    "rawValue": "4"
  },
  "sensor co2": {
    "value": "480.039067269",
    "unit": "ppm",
    "rawValue": "11.6806250763033"
  },
  "spectrometer": {
    "maxFixedIntensity": "16383",
    "integration time in us": "5000",
    "wavelength": Array.<Number>,
    "spectrum": Array.<Number>
  }
}
```

**Comment by Zodiase** [comment]
# Conversion

```
$.environment_sensor_readings.*.timestamp: $.start_time
$.environment_sensor_readings.*.weather_station.airPressure(hPa): $.properties.air_pressure(Pa)
$.environment_sensor_readings.*.weather_station.relHumidity(relHumPerCent): $.properties.relative_humidity(%)
$.environment_sensor_readings.*.weather_station.temperature(DegCelsius): $.properties.air_temperature(K)
$.environment_sensor_readings.*.weather_station.windDirection(degrees) +
$.environment_sensor_readings.*.weather_station.windVelocity(m/s) +
: $.properties.eastward_wind(m/s) + $.properties.northward_wind(m/s)
$.environment_sensor_readings.*.weather_station.precipitation(mm/h): $.properties.precipitation_flux(kg m-2 s-1)
```

**Comment by Zodiase** [comment]
@dlebauer I've listed the conversion I'm sure about. There are a couple other fields I'm not sure about. Could you provide a mapping like you did in https://github.com/terraref/computing-pipeline/issues/156#issuecomment-252131358?

**Comment by dlebauer** [comment]
@Zodiase 

* sunDirection not sure @czender?
* for wind velocity --> if you are converting to north and east vectors make sure to use appropriate trigonometry
* sensor par --> surface_downwelling_photosynthetic_photon_flux_in_air (mol m-2 s-1)
* sensor co2 --> mole_fraction_of_carbon_dioxide_in_air (umol/mol)
  * the units of this can be 'umol / mol' or 'ppm'

For the spectrometer fields, these will be used in the hyperspectral workflow. 
* @FlyingWithJerome and @czender what / where is the conversion from the radiometer values to  radiances as described in https://github.com/terraref/reference-data/issues/30?

**Comment by dlebauer** [comment]
@czender and @FlyingWithJerome where did we leave off with the creation of the Environmental logger files as netcdf? Should we leave those in netcdf files? If so, is there an issue (or should I create a new one) to make those as CF standard names?

**Comment by FlyingWithJerome** [comment]
@dlebauer I doubled checked the [Environmental Logger Script](../blob/master/scripts/environmental_logger/environmental_logger_json2netcdf.py) in the computing pipeline, and the answer to the naming issue is yes. There is a TODO token in the line 252 for standard names.

@Zodiase @czender 
If you would like to know how to convert the units, you may check the [Environmental Logger Script](../blob/master/scripts/environmental_logger/environmental_logger_json2netcdf.py). There is a large global dictionary "_UNIT_DICTIONARY" for you to refer to.

**Comment by czender** [comment]
@Zodiase sunDirection is computed (not measured) from time/location. it is not zenith angle. i'm not sure exactly what it is yet. the sensor documentation does not describe it. and the angle takes values that exceed 90 degrees during daytime, so it is neither a zenith angle nor an altitude. please leave it as sunDirection until we understand it. @solmazhajmohammadi or @TinoDornbusch  do you know how thiesclima defines sunDirection?

**Comment by czender** [comment]
@FlyingWithJerome please note that the environmental logger script was moved to a new repository many months ago: https://github.com/terraref/extractors-environmental
It does not make sense to maintain the version in computing-pipeline. Please work with new repository and deprecate the old.

**Comment by FlyingWithJerome** [comment]
@czender Yes, I had already cloned the new repo

**Comment by rachelshekar** [comment]
New functionality from PiClowder 2 will help with this.  Xingchen will follow up with Max about this.

**Comment by czender** [comment]
@FlyingWithJerome fine but your message above from yesterday points people to the old scripts. please stop doing that. it's confusing.

**Comment by Zodiase** [comment]
@dlebauer @czender
So for `sunDirection`, I'll keep the name, unit and value untouched.
What about `spectrometer`? It's a huge list of numbers in both `spectrometer.wavelength` and `spectrometer.spectrum`.

**Comment by remotesensinglab** [comment]
What sensor we are referring here by spectrometer ? Thanks

**Comment by Zodiase** [comment]
@remotesensinglab I have no idea. But there's model and manufacturer info here https://github.com/terraref/computing-pipeline/issues/252#issuecomment-281804281.

**Comment by czender** [comment]
@Zodiase spectrometer wavelength and spectral fluxes, par, etc. are already archived by the extractor into the netCDF file. Do you want us to modify them in some way? Are you re-writing the environmental logger extractor?

**Comment by Zodiase** [comment]
@czender I think this issue is only to create an extractor to convert data to CF standard and pipe into geostream database.

**Comment by czender** [comment]
Maybe our terminology for "extractor" differs. IMHO "converting to CF standard" should be done in existing extractor (when it is converted from JSON to netCDF). Any new extractor should expect to already find CF-compliant data with which it can do as it wishes.

**Comment by max-zilla** [comment]
@czender @Zodiase so the question basically is, do we:
- expand [this code](https://github.com/terraref/extractors-environmental/tree/master/environmentlogger) to send the variables prepared in [this method](https://github.com/terraref/extractors-environmental/blob/master/environmentlogger/environmental_logger_json2netcdf.py#L195) to send them to geostreams at same time as generation of netCDF
or
- create a new extractor that triggers on netCDF file and extracts variables from there, pushes to geostreams

Broadly speaking, either one of these processes would have to query the dataset metadata to get location, and use that to push a JSON object of properties w/ timestamp to geostreams. 

I personally would be in favor of option 1 as it doesn't result in a new extractor, and we have other extractors like PlantCV that simultaneously generate a CSV and push the CSV contents to BETYdb in one function. This one generating netCDF and push the netCDF variables to Geostreams in one function would be similar. This would be an augmentation to environment logger extractor, not rewriting or replacing per se - sending a JSON blob to Geostreams while we have the vars "in hand" for netCDF creation. We could have that mainProgramTrigger() return a JSON array of properties instead of nothing, for instance.

The main argument for option 2, IMO, is if we expect additional netCDF files it will also handle in addition to the EnvironmentLogger outputs - then it makes sense to separate.

Thoughts?

**Comment by gsrohde** [comment]
@czender Could sunDirection = azimuth?  The sample value shown above of 357.58 (nearly 360) at a little past midnight in a location well to the west of the center of the time zone makes this seem likely.

**Comment by czender** [comment]
I don't have a preference between option 1 and 2. @FlyingWithJerome will soon alter the netCDF file so it's more CF'y (and you can do the same).  If it makes sense to you, you could use the ncks --json feature to pipe the resulting JSON info to Geostreams, either as the last step of the current extractor or as a new extractor.

**Comment by czender** [comment]
@gsrohde it takes two angles to specify the solar position. i agree it's likely that sunDirection is some form of azimuth, but where's the corresponding altitude (complement of zenith angle)? zenith angle (not azimut) is what we need for irradiance purposes. we can compute it using well-known algorithms, but I'd still like to be sure what it is that the JSON file contains.

**Comment by max-zilla** [comment]
@Zodiase let's add a geostreams push to existing EnvironmentLogger extractor output to geostreams, in addition to creating netCDF. 

we can make a separate generic extractor later if we want.

**Comment by FlyingWithJerome** [comment]
@max-zilla @Zodiase 
Let me know if you have any questions on the Environmental logger source code, and I will keep track on your branch.

**Comment by Zodiase** [comment]
I looked at the `extractors-environmental` code and found one (minor-ish?) issue: 

`extractors-meterological` processes the entire dataset in one pass, where it sees all the data and produce mostly full-size aggregated chunks except the head and tail.
To illustrate (`|`s are temporal aggregation cutoffs, `[]`s are final data entries):
```
[...]|[.............]|[.............]|[.............]|[.............]|[.............]|[.............]|[.............]|[.............]|[..........]
// aggregated into 10 entries. 8 of them are full.
```

However `extractors-environmental` processes the dataset one file at a time. With the same aggregation process applied to it, the result would look like:
```
[...]|[........][.....]|[........][.....]|[........][.....]|[........][.....]|[........][.....]|[........][.....]|[........][.....]|[........][.....]|[........][..]
// (extreme case:) aggregated into 19 entries. none of them is full.
```

In other words, since we are working on smaller temporal windows, there will be more fragments created from heads and tails that do not fill a full temporal aggregation block.

**Comment by czender** [comment]
I don't understand the issue that @Zodiase writes about above. @FlyingWithJerome and @Zodiase perhaps you can speak by phone to clarify.

**Comment by Zodiase** [comment]
@czender

During aggregation, for simplicity, all temporal cutoffs are on the minutes with no offset.

Suppose one source data file starts at `09:03:23` and ends on `10:03:23` and we are doing 10-minute aggregation. The first cutoff would be `09:10:00`, so `09:03:23-09:10:00` would be the first aggregated data entry. Whereas the second one would be `09:10:00-09:20:00`. Also the last one would be `10:00:00-10:03:23`. As you can see, the heads and tails are very likely to be very small fragments. In the end you get many non-full-10-minute data entries.

To help with this, `extractors-meterological` waits for the dataset to be filled up and reads the entire dataset so it can combine the source file `09:03:23-10:03:23` with `10:03:23-11:03:23` to eliminate one head fragment and one tail fragment. In this scenario, an entire dataset (24 hours) of data only produces 2 fragments.

**Comment by max-zilla** [comment]
@Zodiase we can change the EnvironmentLogger extractor to wait for a full day of EL files in the dataset before processing - I believe we should still be able to use 23 or 24 as expected file count just like we do with the met datparser. That's probably better anyway so we end up with 1 day netCDF file per EL dataset instead of 24 netCDFs.

@czender we basically have 2 extractors, one takes a day of data + concatenates into one long time series THEN slices, the other takes each hourly file individually and slices that file only starting from the start of the file, so the latter could end up with varying. aggregation bin sizes because our aggregation cutoffs don't divide evenly into the hourly file

**Comment by FlyingWithJerome** [comment]
@Zodiase @max-zilla 

If I understand you correctly, the time period is decided by the raw file passed to EnvironmentLogger, not the EnvironmentLogger iteself. Some of them are collected hourly while the others are collected once two minutes. So for a single day we may have 24 or 720 files

**Comment by max-zilla** [comment]
@FlyingWithJerome if I remember correctly, older datasets will have more files and newer files have the hourly cutoff, but I don't remember where that changed.

**Comment by czender** [comment]
It may be helpful to know that the NCO [ncrcat](http://nco.sf.net/nco.html#ncrcat) operator takes a list of files as input, and concatenates the files while hyperslabbing along the record dimension, aka time, if specified with the -d time,start,end option. This feature can be used to repartition data into files with even hour boundaries, without losing any data. e.g.,
`ncrcat -d time,'2011-03-12 12:00:00,2011-03-12 12:59:59' in*.nc out.nc
`

**Comment by max-zilla** [comment]
@FlyingWithJerome has this been completed? let me know if you have questions for me.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
For sending the data to the geostream database? No, I think I need the host, port and an account of the geostream database to query and insert.

**Comment by max-zilla** [comment]
@FlyingWithJerome i will follow up with you tomorrow. your account will be your Clowder account, the URL is https://terraref.ncsa.illinois.edu/clowder/api/geostreams. can give examples of how to use that endpoint.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
```python
import requests
requests.get("https://terraref.ncsa.illinois.edu/clowder/api/geostreams.").raise_for_status()

# Will return HTTPError: 500 Server Error: Internal Server Error
```

**Comment by max-zilla** [comment]
@FlyingWithJerome ```requests.get("https://terraref.ncsa.illinois.edu/clowder/api/geostreams/datapoints?streamid=300&key=XYZ").raise_for_status()```

I'll email you the key to use instead of XYZ -  keep it secret please.

**Comment by FlyingWithJerome** [comment]
@max-zilla 

To protect the API Key, I plan to write the geostream in a separate module and only upload the pyc/pyo file instead of the source code.

Any comments?

**Comment by dlebauer** [comment]
Please upload source code!

**Comment by max-zilla** [comment]
@FlyingWithJerome upload it without the key included, and pass in the key as a command line argument.

**Comment by robkooper** [comment]
I would prefer to use auth with username/password. That way it gets assigned to a user instead of the system user (which is what happens when you use the key).

**Comment by Zodiase** [comment]
I'm reading the existing environmental logger extractor code. Since I'm not that familiar with Python so I might have missed something, but so far [from this portion](https://github.com/terraref/extractors-environmental/blob/269548581268cec092dd2e5782ed8a848c8d653b/environmentlogger/environmental_logger_json2netcdf.py#L215) I don't think that extractor is doing any unit conversion other than renaming the units to CF standards. For example the conversion from Celsius to Kelvin. When I was working on the meterological extractor I took it as a requirement to convert all values to CF variable names and units suggested in [this table](https://pecan.gitbooks.io/pecan-documentation/content/developers_guide/Adding-an-Input-Converter.html#the-variable-names-should-be-standardname), using "preferred variables" as much as possible (which means breaking `windVelocity` and `windDirection` into `eastward_wind` and `northward_wind`. So far I don't see the environmental logger extractor code doing any of these. Did I miss anything?

**Comment by max-zilla** [comment]
@Zodiase I believe you are correct in both cases - the conversion is for the key only (not the value), and it needs to be both like the met extractor.

**Comment by FlyingWithJerome** [comment]
@max-zilla @Zodiase 
You are correct. Numerical correction on these values will come with geostreaming in the next patch.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
But in this way, the API key will be hard coded in the driver (aka [terra_environmentlogger.py](../blob/master/environmentlogger/terra_environmentlogger.py)?

**Comment by czender** [comment]
@Zodiase thanks for reporting this. A 273 degree temperature bias is worth fixing. @FlyingWithJerome please make sure to convert value not just units. Double-check to make sure there are no missing factors of 10, 100, 1000 missing from the values of any converted quantity. Please note any discrepancies between our data names/standard_names and the table Xingchen points to so we can discuss these Monday.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
The Geostream API works fine yesterday, but today it raises HTTPError 500 again; is it under maintenance right now?

**Comment by max-zilla** [comment]
@FlyingWithJerome the postgres service was restarted, Clowder needed to be restarted as well to reconnect. Should work now.

**Comment by FlyingWithJerome** [comment]
@max-zilla 

I double checked the API documentation in the clowder and cannot find the parts for Geostream. I would like to know the the host address I can POST to, and whether it supports JSON-Encoded data (seems that it does not support Python json.dumps() encoder).

**Comment by max-zilla** [comment]
@FlyingWithJerome here are two examples of extractors that post to Geostreams:
https://github.com/terraref/extractors-meterological/blob/master/datparser/terra_met_datparser.py#L244
https://github.com/terraref/extractors-metadata/blob/master/sensorposition/terra_sensorposition.py#L162

See create_sensor, get_sensor_id, create_stream, get_stream_id also.

Basically for geostreams...
1. "sensor" is top level - ignore the word sensor because it's confusing (from another project). For TERRA, each "sensor" is a plot in the field.
2. "stream" is each instrument x each plot. so if Plot="Range 1 Pass 1" then a stream might be "VNIR - Range 1 Pass 1".
3. "datapoint" is a single measurement at a single point in time, e.g. reflectance. these go into a stream.

Big idea is we have a field of plots (1) and each plot has a list of instruments (2) that each have a bunch of datapoints (3) indicating what that instrument captured in that plot.

**Comment by FlyingWithJerome** [comment]
@max-zilla 

Thanks, these two examples make more sense to me. I will double check if I have the pyclowder dependencies on my home machines and do some little researches on them

**Comment by max-zilla** [comment]
@FlyingWithJerome make sure you use the newer PyClowder 2: https://opensource.ncsa.illinois.edu/bitbucket/projects/CATS/repos/pyclowder2/browse
```
python setup.py install --force
```

The --force will overwrite any pyclowder 1 files if you already had them installed.

**Comment by dlebauer** [comment]
@FlyingWithJerome there is some confusion on this side - have you taken on this issue? If so, that is great. Or is there anything that @Zodiase can help with?

One other thought - instead of creating one file per day, can this netcdf file just be appended in time, perhaps creating a new one each year?

**Comment by FlyingWithJerome** [comment]
@dlebauer
The unit conversion issue had already been fixed; I'm right now polishing the code based on Craig's feedback and adding geostream to the pipeline. Max had provided me some useful references on Geostream API.

**Comment by rachelshekar** [comment]
@FlyingWithJerome  - can you give an update on this issue?  Can it be closed?

**Comment by FlyingWithJerome** [comment]
@max-zilla 
Hi Max, I would like to know how to fill in the  "geom," "type," and "region" for create_sensor(), since the Environmental Logger has no region, type or geometry. If I had all of them as None, I will get a 400 Bad Request.
BTW, "connector" can be just a random Connector instance?

**Comment by FlyingWithJerome** [comment]
@rachelshekar I'm working on this; will be finished in a not-so-far future

**Comment by max-zilla** [comment]
@FlyingWithJerome for the met extractor, we are using a location off to the side of the field (on the building nearby) to represent "full field" data:

```
def process_message(self, connector, host, secret_key, resource, parameters):
self.sensor_name = "Full Field"
main_coords = [ -111.974304, 33.075576, 0]

sensor_id = get_sensor_id(host, secret_key, self.sensor_name)
		if not sensor_id:
			sensor_id = create_sensor(host, secret_key, self.sensor_name, {
				"type": "Point",
				# These are a point off to the right of the field
				"coordinates": main_coords
			})


stream_name = self.sensor_name + " - Weather Station"
		stream_id = get_stream_id(host, secret_key, stream_name)
		if not stream_id:
			stream_id = create_stream(host, secret_key, sensor_id, stream_name, {
				"type": "Point",
				"coordinates": main_coords
			})
```

I would replace "Weather Station" with EnvironmentLogger but otherwise you can use the same Full Field sensor we're using for the weather data.

The connector is passed into the process_message method for you.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
In PyClowder2 the create_sensor() takes 7 arguments, but you only pass 4?

**Comment by max-zilla** [comment]
@FlyingWithJerome ah yes, didn't realize the example function was custom before pyClowder 2 had geostreams.

```
body = {
		"name": name,
		"type": "point",
		"geometry": {
				"type": "Point",
				# These are a point off to the right of the field
				"coordinates": main_coords
			},
		"properties": {
			"popupContent": name,
			"type": {
				"id": "MAC Field Scanner",
				"title": "MAC Field Scanner",
				"sensorType": 4
			},
			"name": name,
			"region": "Maricopa"
		}
	}
```

So for using PyC2:

geom = {
				"type": "Point",
				# These are a point off to the right of the field
				"coordinates": main_coords
			}

type = {
				"id": "MAC Field Scanner",
				"title": "MAC Field Scanner",
				"sensorType": 4
			}

region = "Maricopa"

**Comment by FlyingWithJerome** [comment]
@max-zilla 
These are extremely helpful! Thanks!

**Comment by dlebauer** [comment]
Just to be clear - the EnvirontmentLogger and WeatherStation should be listed as distinct locations (sensors). 

* The WeatherStation coordinates are 33.074457 N, 111.975163 W https://github.com/terraref/computing-pipeline/issues/252#issuecomment-279223800 
* The EnvironmentLogger is on top of the gantry system so it is moving around. For the EnvironmentLogger I think it would be better to have a polygon that circumscribes the gantry than to have a point off to the side (though for mapping purposes it might be preferable to keep it off to the side).

**Comment by max-zilla** [comment]
@dlebauer noted. we can change that location/geometry at any time fortunately, will make a note.

**Comment by FlyingWithJerome** [comment]
@dlebauer 
I think the EL records have the weather station data, but has no position data (i.e., seems that we cannot trace the EL Logger position like what we do on hyperspectral cameras). Only the weather station in EL records is in a fixed position, and every sensors else (sensor par, sensor co2, etc.) are movable?

**Comment by max-zilla** [comment]
@FlyingWithJerome i think that's correct - that's why David suggests using a polygon to represent the entire field or something else off to the side if there are overlap issues. We can experiment with that a bit later.

**Comment by FlyingWithJerome** [comment]
@max-zilla 
Sounds like a plan. I will finish the geostream stuff these days and send to David and you for review

**Comment by rachelshekar** [comment]
@FlyingWithJerome can this issue be closed?

**Comment by max-zilla** [comment]
@rachelshekar yes I merged @FlyingWithJerome 's pull request.
